noir/aztec_macros/src/lib.rs

```diff
@@ -1601,3 +1601,143 @@ fn event_signature(event: &StructType) -> String {
+        return Ok(());
+    }
+
+    // In order to implement compute_note_hash_and_nullifier, we need to know all of the different note types the
+    // contract might use. These are the types that implement the NoteInterface trait, which provides the
+    // get_note_type_id function.
+
+    let mut note_types: Vec<String> = Vec::new();
+
+    // These note types can be declared in either external crates, or the contract's own crate. External crates have
+    // already been processed and resolved, but are available here via the NodeInterner. This is because the contract's
+    // crate is the last crate to be processed, after all of its dependencies.
+    for trait_impl_id in 0..(&context.def_interner.next_trait_impl_id()).0 {
+        let trait_impl = &context.def_interner.get_trait_implementation(TraitImplId(trait_impl_id));
+
+        if trait_impl.borrow().ident.0.contents == "NoteInterface" {
+            if let Type::Struct(s, _) = &trait_impl.borrow().typ {
+                note_types.push(s.borrow().name.0.contents.clone());
+            } else {
+                panic!("Found impl for NoteInterface on non-Struct");
+            }
+        }
+    }
+
+    // This crate's traits and impls have not yet been resolved, so we look for impls in unresolved_trait_impls.
+    note_types.extend(unresolved_traits_impls.iter()
+        .filter(|trait_impl| trait_impl.trait_path.segments.last().unwrap().0.contents == "NoteInterface")
+        .filter_map(|trait_impl| match &trait_impl.object_type.typ {
+            UnresolvedTypeData::Named(path, _, _) => Some(path.segments.last().unwrap().0.contents.clone()),
+            _ => None,
+        }));
+
+    // We can now generate a version of compute_note_hash_and_nullifier tailored for the contract in this crate.
+    let func = generate_compute_note_hash_and_nullifier(&note_types);
+
+    // And inject the newly created function into the contract.
+
+    // We don't have a reasonable location for the source code of this autogenerated function, so we simply pass an
+    // empty span. This function should not produce errors anyway so this should not matter.
+    let location = Location::new(Span::empty(0), file_id);
+
+    // These are the same things the ModCollector does when collecting functions: we push the function to the
+    // NodeInterner, declare it in the module (which checks for duplicate definitions), and finally add it to the list
+    // on collected but unresolved functions.
+
+    let func_id = context.def_interner.push_empty_fn();
+    context.def_interner.push_function(func_id, &func.def, ModuleId { krate: *crate_id, local_id: module_id }, location);
+
+    let Ok(_) = context.def_map_mut(crate_id).unwrap()
+        .modules_mut()[module_id.0]
+        .declare_function(
+            func.name_ident().clone(), func_id
+        )
+    else {
```

Comment:
This entire let Ok(_) block can just be an `.declare_functionunwrap_or_else(|| panic!("Failed to ... "))`